pkg/repository/remote/upload.go

```diff
@@ -0,0 +1,213 @@
+	"os"
+	"runtime"
+	"strings"
+	"time"
+
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/pingcap-incubator/tiup/pkg/repository/v1manifest"
+	"github.com/pingcap-incubator/tiup/pkg/utils"
+	"github.com/pingcap-incubator/tiup/pkg/version"
+)
+
+// Transporter defines methods to upload components
+type Transporter interface {
+	WithOS(os string) Transporter
+	WithArch(arch string) Transporter
+	WithDesc(desc string) Transporter
+	Open(tarbal string) error
+	Close() error
+	Upload() error
+	Sign(key *v1manifest.KeyInfo, m *v1manifest.Component) error
+}
+
+type transporter struct {
+	tarFile     *os.File
+	tarInfo     os.FileInfo
+	sha256      string
+	os          string
+	arch        string
+	entry       string
+	component   string
+	version     string
+	description string
+	endpoint    string
+}
+
+// New returns a Transporter
+func New(endpoint, component, version, entry string) Transporter {
+	return &transporter{
+		endpoint:  endpoint,
+		component: component,
+		entry:     entry,
+		os:        runtime.GOOS,
+		arch:      runtime.GOARCH,
+		version:   version,
+	}
+}
+
+func (t *transporter) WithOS(os string) Transporter {
+	t.os = os
+	return t
+}
+
+func (t *transporter) WithDesc(desc string) Transporter {
+	t.description = desc
+	return t
+}
+
+func (t *transporter) WithArch(arch string) Transporter {
+	t.arch = arch
+	return t
+}
+
+func (t *transporter) Open(tarbal string) error {
+	info, err := os.Stat(tarbal)
+	if err != nil {
+		return err
+	}
+	t.tarInfo = info
+
+	file, err := os.Open(tarbal)
+	if err != nil {
+		return err
+	}
+	t.tarFile = file
+
+	sha256, err := utils.SHA256(file)
+	if err != nil {
+		file.Close()
+		return err
+	}
+	t.sha256 = sha256
+
+	if _, err := file.Seek(0, io.SeekStart); err != nil {
+		file.Close()
+		return err
+	}
+
+	return nil
+}
+
+func (t *transporter) Close() error {
+	return t.tarFile.Close()
+}
+
+func (t *transporter) Upload() error {
+	postAddr := fmt.Sprintf("%s/api/v1/tarball/%s", t.endpoint, t.sha256)
+	tarbalName := fmt.Sprintf("%s-%s-%s-%s.tar.gz", t.component, t.version, t.os, t.arch)
+	resp, err := utils.PostFile(t.tarFile, postAddr, "file", tarbalName)
+	if err != nil {
+		return err
+	}
+	return resp.Body.Close()
+}
+
+func (t *transporter) Sign(key *v1manifest.KeyInfo, m *v1manifest.Component) error {
+	id, err := key.ID()
+	if err != nil {
+		return err
+	}
+
+	if m == nil {
+		m = t.defaultComponent()
+	} else {
+		m.Version++
+	}
+
+	if strings.Contains(t.version, version.NightlyVersion) {
+		m.Nightly = t.version
+	}
+
+	platformStr := fmt.Sprintf("%s/%s", t.os, t.arch)
+	if m.Platforms[platformStr] == nil {
+		m.Platforms[platformStr] = map[string]v1manifest.VersionItem{}
+	}
+	m.Platforms[platformStr][t.version] = v1manifest.VersionItem{
+		Entry:    t.entry,
+		Released: time.Now().Format(time.RFC3339),
+		URL:      fmt.Sprintf("/%s-%s-%s-%s.tar.gz", t.component, t.version, t.os, t.arch),
+		FileHash: v1manifest.FileHash{
+			Hashes: map[string]string{
```

Comment:
We generate both SHA256 and SHA512 checksums during migrate, maybe add them both? If use only one of them, I'd prefer SHA512.